libgin/annex/util.go

```diff
@@ -1,12 +1,136 @@
 package annex
 
-import "github.com/G-Node/git-module"
+import (
+	"bufio"
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
 
-func isAnnexed(dir string) (bool, error) {
-	return false, nil
+	"github.com/gogs/git-module"
+)
+
+const annexDirLetters = "0123456789zqjxkmvwgpfZQJXKMVWGPF"
+
+func IsAnnexFile(blob *git.Blob) bool {
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	blob.DataPipeline(stdout, stderr)
+	reader := bufio.NewReader(stdout)
+
+	// Contents differ if the file is locked (symlink) or unlocked (plain text blob)
+	// symlinks contain a relative path to the annex content store (.git/annex/objects)
+	// unlocked pointers contain a path rooted at /annex/objects
+	// It's important not to look through the entire file for the string '/annex/objects'
+	// for a couple of reasons:
+	// 1. The file might be very large
+	// 2. The file might contain the string for other reasons (e.g., documentation, or this file)
+
+	if blob.IsLink() {
+		// read the entire contents and check if it's a path inside .git/annex/objects
+		readbuf := make([]byte, reader.Size())
+		_, err := reader.Read(readbuf)
+		if err != nil {
+			fmt.Printf("ERROR: %s\n", err.Error())
+		}
+		return strings.Contains(string(readbuf), ".git/annex/objects")
+	}
+
+	// check if the first bytes match /annex/objects
+	const ident = "/annex/objects"
+	readbuf := make([]byte, len(ident))
+	_, err := reader.Read(readbuf)
+	if err != nil {
+		fmt.Printf("ERROR: %s\n", err.Error())
+	}
+	return string(readbuf) == ident
 }
 
 func Upgrade(dir string) (string, error) {
-	cmd := git.NewACommand("upgrade")
+	cmd := git.NewCommand("annex", "upgrade")
 	return cmd.RunInDir(dir)
 }
+
+func IsBare(repo *git.Repository) (bool, error) {
+	out, err := git.NewCommand("config", "core.bare").RunInDir(repo.Path)
+	out = strings.TrimSpace(out)
+	if err != nil {
+		return false, err
+	}
+	if out == "true" {
+		return true, nil
+	}
+	if out == "false" {
+		return false, nil
+	}
+	return false, fmt.Errorf("unexpected output: %s", out)
+}
+
+// ContentLocation returns the location of the content file for a given annex key.
+// The returned path is relative to the repository git directory.
+func ContentLocation(repo *git.Repository, key string) (string, error) {
+	gitdir := repo.Path
+	if bare, err := IsBare(repo); err != nil {
+		return "", err
+	} else if !bare {
+		gitdir = filepath.Join(gitdir, ".git")
+	}
+	objectstore := filepath.Join(gitdir, "annex", "objects")
+	// there are two possible object paths depending on annex version
+	// the most common one is the newest, but we should try both anyway
+	objectpath := filepath.Join(objectstore, hashdirmixed(key), key)
+	if _, err := os.Stat(objectpath); os.IsNotExist(err) {
```

Comment:
Is it possible that an error occurs at this point that is not an `NotExist` error; if yes, should this be handled?

Reply:
There's not much info about what other kind of errors can occur, but we should cover all bases.  Even if it's currently the only type of error, it might not be in the future.